### PR TITLE
Tail EMA Checkpoint Averaging: Average Late-Training EMA Snapshots

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

**Tail Checkpoint Averaging** averages EMA model states from the last 20% of training. This is fundamentally different from:
- **SWAD (#2094, catastrophic):** averaged raw model along full trajectory — captured optimization noise
- **Model soup (#2142, catastrophic):** averaged across independent seeds — loss barriers between basins
- **This:** averages the **EMA model** (already smoothed) from the **converged tail** (same basin, same seed)

Late-training EMA snapshots sit in the same loss basin and near the same local minimum. Their average occupies the centroid of the converged region — empirically flatter and more robust to OOD shifts. This is standard practice in Kaggle competitions and NLP (checkpoint averaging is used by virtually all top competition solutions and transformer training pipelines).

**Why it should work where SWAD didn't:** SWAD averaged over the FULL trajectory, including unstable early phases and high-LR exploration. Tail averaging only touches the stable, converged portion where the model has settled. The EMA model is already smoothed, so averaging EMA snapshots is doubly smoothed — very conservative.

**Expected impact:** p_tan -0.5% to -2%. Zero additional training cost — just save more checkpoints.

### Step 1: Add checkpoint saving at multiple epochs

In the training loop, save the EMA model state at evenly-spaced epochs in the tail:

```python
# In Config:
tail_avg_start_epoch: int = 0  # 0 = disabled. Set to e.g. 135 to save EMA snapshots from epoch 135 onward
tail_avg_interval: int = 5     # Save EMA every N epochs after tail_avg_start_epoch

# In argparse:
parser.add_argument('--tail_avg_start_epoch', type=int, default=0)
parser.add_argument('--tail_avg_interval', type=int, default=5)
```

In the training loop, after the EMA update step:
```python
# After EMA update (around line ~1935), add:
if cfg.tail_avg_start_epoch > 0 and ema_model is not None and epoch >= cfg.tail_avg_start_epoch:
    if (epoch - cfg.tail_avg_start_epoch) % cfg.tail_avg_interval == 0:
        _tail_ckpts.append(copy.deepcopy(ema_model.state_dict()))
        print(f"[tail_avg] Saved EMA snapshot at epoch {epoch} (total: {len(_tail_ckpts)})")
```

Initialize `_tail_ckpts = []` before the training loop. Ensure deepcopy sends to CPU to avoid GPU memory pressure:
```python
_tail_ckpts.append({k: v.cpu().clone() for k, v in ema_model.state_dict().items()})
```

### Step 2: Average tail checkpoints after training

After training completes (before final evaluation):
```python
if cfg.tail_avg_start_epoch > 0 and len(_tail_ckpts) >= 2:
    print(f"[tail_avg] Averaging {len(_tail_ckpts)} EMA snapshots")
    avg_state = copy.deepcopy(_tail_ckpts[0])
    for key in avg_state:
        if avg_state[key].is_floating_point():
            avg_state[key] = sum(c[key] for c in _tail_ckpts) / len(_tail_ckpts)
    # Move back to GPU
    avg_state = {k: v.to(next(ema_model.parameters()).device) for k, v in avg_state.items()}
    ema_model.load_state_dict(avg_state)
    print(f"[tail_avg] Loaded averaged EMA model for evaluation")
```

This replaces the final EMA model with the averaged version before the standard evaluation loop runs. No changes to the eval code needed.

### Step 3: Run experiments — 2 seeds x 2 configs = 4 runs

**Config A: Save from epoch 135 every 5 epochs** (captures ~4-5 snapshots: 135, 140, 145, 150, 155)
```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent frieren \
  --wandb_name "frieren/tail-avg-135-s42" --wandb_group phase6/tail-ema-avg \
  --tail_avg_start_epoch 135 --tail_avg_interval 5 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Seed 73 (same, --seed 73, wandb_name "frieren/tail-avg-135-s73")
```

**Config B: Save from epoch 145 every 3 epochs** (captures ~4 snapshots: 145, 148, 151, 154)
```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent frieren \
  --wandb_name "frieren/tail-avg-145-s42" --wandb_group phase6/tail-ema-avg \
  --tail_avg_start_epoch 145 --tail_avg_interval 3 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Seed 73 (same, --seed 73, wandb_name "frieren/tail-avg-145-s73")
```

### What to report

| Config | Seed | p_in | p_oodc | p_tan | p_re | W&B | # snapshots averaged |
|--------|------|------|--------|-------|------|-----|---------------------|
| tail_avg from 135 | 42 | | | | | | |
| tail_avg from 135 | 73 | | | | | | |
| tail_avg from 145 | 42 | | | | | | |
| tail_avg from 145 | 73 | | | | | | |
| **Best config avg** | — | | | | | | |
| **Baseline (no tail avg)** | — | 13.05 | 7.70 | 28.60 | 6.55 | d7l91p0x, j9btfx09 | 1 (final only) |

Key comparison: tail-averaged EMA vs standard single-checkpoint EMA. Also report: print the individual EMA snapshot val losses (if available) to see the trend in late-training EMA quality.

### Memory note
Each EMA state deepcopy adds ~100-150 MB to CPU memory. With 5 snapshots that's <1 GB — well within budget.

## Baseline

Current baseline (PR #2130, GSB + PCGrad):

| Metric | 2-seed target |
|--------|--------------|
| p_in   | < 13.05      |
| p_oodc | < 7.70       |
| p_tan  | **< 28.60**  |
| p_re   | < 6.55       |

W&B: d7l91p0x (s42, p_tan=28.9), j9btfx09 (s73, p_tan=28.3)

```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-gsb-pcgrad" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Summary

Tail EMA checkpoint averaging had **negligible effect** on all metrics — differences are within noise (±0.3%). The hypothesis did not produce the expected -0.5% to -2% improvement on p_tan.

### Key Issue: EMA starts at epoch 140

The `ema_start_epoch=140` means `ema_model is None` before epoch 140. Config A (start=135) effectively started at epoch 140, not 135. Config B (start=145) worked as intended. Both configs captured fewer snapshots than planned:
- Config A: 3 snapshots (epochs 140, 145, 150) instead of 4 (135, 140, 145, 150)
- Config B: 2 snapshots (epochs 145, 148) instead of 3+ (145, 148, 151)

### Metrics: Standard EMA vs Tail-Averaged EMA

**Baseline:** p_tan < 28.60, p_in < 13.05, p_oodc < 7.70, p_re < 6.55

| Run | Snaps | Standard EMA p_tan | Tail-Avg p_tan | Δ |
|-----|-------|-------------------|----------------|---|
| Config A (135/5), seed 42 | 3 | 29.0 | 29.11 | +0.4% |
| Config A (135/5), seed 73 | 2 | 29.2 | 29.16 | -0.1% |
| Config B (145/3), seed 42 | 2 | 29.8 | 29.85 | +0.2% |
| Config B (145/3), seed 73 | 2 | 29.7 | 29.62 | -0.3% |

**Full surface pressure breakdown (standard EMA → tail-avg):**

| Run | p_in | p_tan | p_oodc | p_re |
|-----|------|-------|--------|------|
| A s42: EMA | 13.0 | 29.0 | 7.8 | 6.6 |
| A s42: tail | 12.95 | 29.11 | 7.91 | 6.67 |
| A s73: EMA | 13.9 | 29.2 | 7.7 | 6.5 |
| A s73: tail | 14.06 | 29.16 | 7.93 | 6.55 |
| B s42: EMA | 13.5 | 29.8 | 8.0 | 6.5 |
| B s42: tail | 13.44 | 29.85 | 8.01 | 6.43 |
| B s73: EMA | 14.0 | 29.7 | 8.0 | 6.7 |
| B s73: tail | 14.07 | 29.62 | 8.04 | 6.70 |

None of the runs beat the baseline p_tan < 28.60 (best standard EMA: 29.0, best tail-avg: 29.11).

### Commands

```bash
# Config A (start=135, interval=5)
CUDA_VISIBLE_DEVICES=0 python train.py --agent frieren \
  --wandb_name "frieren/tail-avg-135-s42" --wandb_group phase6/tail-ema-avg \
  --tail_avg_start_epoch 135 --tail_avg_interval 5 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Config B: same but --tail_avg_start_epoch 145 --tail_avg_interval 3
```

### W&B Run IDs
- Config A, seed 42: `fh7f24u6` (3 snapshots, 181.4 min)
- Config A, seed 73: `azir19g1` (2 snapshots, 180.6 min)
- Config B, seed 42: `wrh72afu` (2 snapshots, 181.2 min)
- Config B, seed 73: `6hhy1mih` (2 snapshots, 181.4 min)

Peak memory: 44.8-46.2 GB

### What happened

The tail-averaged EMA model performed essentially identically to the standard best-epoch EMA checkpoint across all 4 experiments. Differences ranged from -0.3% to +0.4% on p_tan — well within run-to-run noise.

**Why it didn't work:**
1. **EMA is already extremely smooth.** With `ema_decay=0.999`, the EMA model is already a heavily smoothed average over ~1000 gradient steps. Averaging 2-3 EMA snapshots taken 5 epochs apart adds very little additional smoothing — the EMA at epoch 150 already "remembers" epochs 140-149 through its exponential weighting.
2. **Too few snapshots.** Only 2-3 snapshots were averaged, limiting the potential smoothing effect. The EMA start at epoch 140 reduced Config A from 4→3 snapshots.
3. **Converged EMA snapshots are nearly identical.** In the final 10-15 epochs of training with a cosine LR schedule (T_max=160), the learning rate is very small. The EMA model barely changes between snapshots, so averaging them produces nearly the same model.

### Suggested follow-ups

1. **Lower EMA decay for tail averaging:** Use `ema_decay=0.99` (instead of 0.999) to create more diverse snapshots, then average them. The individual snapshots would be noisier but their average might find a flatter minimum.
2. **Start tail averaging much earlier:** Try start_epoch=100 with interval=10 to capture snapshots across a wider training phase (still post-convergence but with more LR diversity).
3. **Prediction-level averaging (ensemble):** Instead of weight averaging, run inference with multiple checkpoints and average their predictions. This works even when models are in different basins and avoids the weight-space interpolation problem entirely.
4. **Stochastic Weight Averaging (SWA) with cyclic LR:** Use a cyclic learning rate in the tail to explore a wider region of the loss basin, then average the resulting checkpoints. This is different from naive tail averaging because the cyclic LR creates genuinely diverse snapshots.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2146 because ADVISOR: Closing — tail EMA averaging produces negligible difference (±0.3%, pure noise). Clean null result.

Your analysis nails it: with ema_decay=0.999, the EMA already averages over ~1000 gradient steps. Additional averaging of 2-3 late snapshots is redundant. The EMA barely changes between epochs 140-155 because the cosine LR is near zero.

This confirms: **post-hoc weight averaging approaches are exhausted for this setup.** SWAD, model soup, and tail averaging all fail for different but related reasons. The EMA + cosine schedule is already a highly effective averaging mechanism.

New assignment coming shortly..
